lib/timers.js

```diff
@@ -54,99 +112,132 @@ exports.active = function(item) {
-          process.domain = null;
-          process.nextTick(listOnTimeoutNT, list);
-          process.domain = oldDomain;
-        }
-      }
+      domain.enter();
     }
+
+    _tryOnTimeout(timer, list);
+
+    if (domain)
+      domain.exit();
   }
 
+  // If `L.peek(list)` returned nothing, the list was either empty or we have
+  // called all of the timer timeouts.
+  // As such, we can remove the list and clean up the TimerWrap C++ handle.
   debug('%d list empty', msecs);
   assert(L.isEmpty(list));
-  list.close();
-  delete lists[msecs];
+  this.close();
+  if (list._unrefed === true) {
+    delete unrefedLists[msecs];
+  } else {
+    delete refedLists[msecs];
+  }
+}
+
+
+// An optimization so that the try/finally only de-optimizes what is in this
+// smaller function.
+function _tryOnTimeout(timer, list) {
+  timer._called = true;
+  var threw = true;
+  try {
+    timer._onTimeout();
+    threw = false;
+  } finally {
+    if (!threw) return;
+
+    // We need to continue processing after domain error handling
+    // is complete, but not by using whatever domain was left over
+    // when the timeout threw its exception.
+    const domain = process.domain;
+    process.domain = null;
+    // If we threw, we need to process the rest of the list in nextTick.
+    process.nextTick(listOnTimeoutNT, list);
+    process.domain = domain;
+  }
 }
 
 
 function listOnTimeoutNT(list) {
-  list[kOnTimeout]();
+  list._timer[kOnTimeout]();
 }
 
 
+// A convenience function for re-using TimerWrap handles more easily.
+//
+// This mostly exists to fix https://github.com/nodejs/node/issues/1264.
+// Handles in libuv take at least one `uv_run` to be registered as unreferenced.
+// Re-using an existing handle allows us to skip that, so that a second `uv_run`
+// will return no active handles, even when running `setTimeout(fn).unref()`.
 function reuse(item) {
   L.remove(item);
 
-  var list = lists[item._idleTimeout];
+  var list = refedLists[item._idleTimeout];
   // if empty - reuse the watcher
   if (list && L.isEmpty(list)) {
     debug('reuse hit');
-    list.stop();
-    delete lists[item._idleTimeout];
-    return list;
+    list._timer.stop();
+    delete refedLists[item._idleTimeout];
+    return list._timer;
   }
 
   return null;
 }
 
 
+// Remove a timer. Cancels the timeout and resets the relevant timer properties.
 const unenroll = exports.unenroll = function(item) {
-  var list = reuse(item);
-  if (list) {
+  var handle = reuse(item);
+  if (handle) {
     debug('unenroll: list empty');
-    list.close();
+    handle.close();
   }
   // if active is called later, then we want to make sure not to insert again
   item._idleTimeout = -1;
 };
 
 
-// Does not start the time, just sets up the members needed.
+// Make a regular object able to act as a timer by setting some properties.
+// This function does not start the timer, see `active()`.
+// Using existing objects as timers slightly reduces object overhead.
```

Comment:
Ah OK, it wasn't clear to me from the comments, but it might be clear for others. I would suggest rephrasing that as follows:

> Sets additional timer-specific properties on the object "item" itself rather than on a separate sub-object to avoid the overhead of an extra level of property access.